### PR TITLE
Performance Metrics and Improvements

### PR DESCRIPTION
**Pull request by azenla**
Some very important things that were added in this change are metrics to measure the number of open requests, the number of subscriptions, and the number of cached nodes.

### CODE REVIEW

File: lib/src/entry/ongoing_query.dart [hunk 1]
```diff
@@ -15,87 +15,89 @@ class OngoingQuery {
...
-          });
         }
-      }
 
-      if (forceRefresh) {
         forEachResponse((response) {
-          sendToResponse(response);
+          response.updateStream(
+            rows.values.where((r) {
+              return r.id >= row.id;
```
Comment (Pacane) [inline lib/src/entry/ongoing_query.dart]: could be a 1 liner.

File: lib/src/entry/ongoing_query.dart [hunk 2]
```diff
@@ -15,87 +15,89 @@ class OngoingQuery {
...
+          );
+        });
+      } else {
+        QueryTableRow row = rows[path];
+        row.values.addAll(update.values);
+        forEachResponse((response) {
+          response.updateStream(
+            [row.format(knownColumns)],
+            meta: {
+              "modify": "replace ${row.id}-${row.id}"
```
Comment (Pacane) [inline lib/src/entry/ongoing_query.dart:88]: is that `"replace ${row.id}-${row.id}"` correct? Kinda looks like a copy-paste mistake, I'm not sure what this means anyway.

File: lib/src/process/stream.dart [hunk 3]
```diff
@@ -59,6 +59,7 @@ abstract class QueryStream extends Stream<QueryUpdate> {
   }
 
   @override
+  // ignore: invalid_method_override_return_type
```
Comment (Pacane) [inline lib/src/process/stream.dart:62]: any reason for that ignore? Shouldn't we have `WrappedQueryStream` implement `QueryStream` or something like that?

**Review by Pacane — CHANGES_REQUESTED** [review]

File: lib/src/entry/ongoing_query.dart (hunk 2 above)
Comment (azenla) [reply lib/src/entry/ongoing_query.dart:88]: Replace supports ranges, so this is correct. It tells it to replace only one row.

File: lib/src/process/stream.dart (hunk 3 above)
Comment (azenla) [reply lib/src/process/stream.dart:62]: Generic type problems. I'm not entirely sure why it does that.

File: benchmark/filter_parser.dart [hunk 4]
```diff
@@ -0,0 +1,81 @@
+import "package:dql/parsers/filter.dart";
+
+final List<String> inputs = <String>[];
+final List<String> slows = <String>[];
```
Comment (Pacane) [inline benchmark/filter_parser.dart:4]: what's this for?

File: lib/parsers/query.dart [hunk 5]
```diff
@@ -165,27 +175,44 @@ PathExpression parseExpressionInput(String input) {
...
-      count++;
-      starCount++;
-      return r".*";
+
+  var sections = input.split("/");
+
+  var hasHitModifier = false;
+  var topmost = "";
+
+  var patternPartials = sections.map((section) {
```
Comment (Pacane) [inline lib/parsers/query.dart:184]: I feel like so many things in there could be extracted. It's really hard to read/understand.

File: lib/parsers/query.dart [hunk 6]
```diff
@@ -165,27 +175,44 @@ PathExpression parseExpressionInput(String input) {
...
     }
 
-    return match.group(0);
-  }, onNonMatch: (String str) {
-    return escapeRegex(str);
-  });
+    return result;
+  }).toList();
+
+  String ptrn = patternPartials.join("/");
```
Comment (Pacane) [inline lib/parsers/query.dart:214]: ptrn => pattern, no need for these abbreviations.

File: lib/src/commands/list.dart [hunk 7]
```diff
@@ -226,7 +231,7 @@ class ListNodeQueryProcessor extends QueryProcessor {
             isSubBroker = false;
           }
 
-          if (isSubBroker && traverseBrokers == false) {
+          if (isSubBroker && traverseBrokers == false && !(expression.topmost == path)) {
```
Comment (Pacane) [inline lib/src/commands/list.dart:234]: !(x == y) ==> x != y

File: lib/src/commands/list.dart [hunk 8]
```diff
@@ -44,6 +45,10 @@ class ListNodeQueryProcessor extends QueryProcessor {
       enableActions = true;
     }
 
+    if (stream.getAttribute("option:enableCache") == false) {
```
Comment (Pacane) [inline lib/src/commands/list.dart:48]: I think these could be initialized with ternary operators and turned into `final`

File: lib/src/commands/sublist.dart [hunk 9]
```diff
@@ -3,19 +3,42 @@ part of dslink.dql.query;
...
-    return _handles;
+    if (controller != null) {
+      controller.close();
+      controller = null;
+    }
+
+    return _handles.keys;
+  }
+
+  void onOldUpdate() {
```
Comment (Pacane) [inline lib/src/commands/sublist.dart]: I'm not sure I understand what's the deal with the "oldUpdate" thing.. Care to explain?

File: lib/src/commands/sublist.dart [hunk 10]
```diff
@@ -112,17 +141,20 @@ class SublistQueryProcessor extends QueryProcessor {
...
-        holders[path] = holder;
+          holder.listSub = holder.create(this).listen((QueryUpdate update) {
+            controller.add(update);
+          });
+
+          holder.controller = controller;
+          holders[path] = holder;
+        }
+        holder.oldUpdate = update;
+        holder.onOldUpdate();
```
Comment (Pacane) [inline lib/src/commands/sublist.dart]: I'd personally pass the update to `onOldUpdate` and assign it in there. If it makes sense.

**Review by Pacane — CHANGES_REQUESTED** [review]

File: lib/src/commands/sublist.dart (hunk 9 above)
Comment (azenla) [reply lib/src/commands/sublist.dart]: We have a stream of table updates, and we are interjecting that update into new updates that match.

```
path | something
abc   | 123
```

=>

```
path | something
abc/1 | 123
abc/2 | 123
abc/3 | 123
```

File: lib/src/commands/list.dart (hunk 8 above)
Comment (azenla) [reply lib/src/commands/list.dart:48]: These attributes can change. But I'll give it a look.

File: lib/src/commands/list.dart (hunk 7 above)
Comment (azenla) [reply lib/src/commands/list.dart:234]: Didn't even notice that.

File: lib/parsers/query.dart (hunk 6 above)
Comment (azenla) [reply lib/parsers/query.dart:214]: Agreed, I think that might be an identifier conflict, but not sure.

File: lib/parsers/query.dart (hunk 5 above)
Comment (azenla) [reply lib/parsers/query.dart:184]: Potentially. It's a complicated parsing technique.

File: benchmark/filter_parser.dart (hunk 4 above)
Comment (azenla) [reply benchmark/filter_parser.dart:4]: I think I intended this to track slow queries and warn about them.

File: lib/src/commands/sublist.dart (hunk 10 above)
Comment (azenla) [reply lib/src/commands/sublist.dart]: I could do that, I'll change it.

File: benchmark/filter_parser.dart (hunk 4 above)
Comment (Pacane) [reply benchmark/filter_parser.dart:4]: Doesn't seem to be used after all.

**Pull request closed without merging**